pkg/upgrade/upgrades/system_users_role_id_migration.go

```diff
@@ -0,0 +1,173 @@
+ALTER TABLE system.users ALTER COLUMN "user_id" SET NOT NULL
+`
+
+const addUserIDIndex = `
+CREATE UNIQUE INDEX users_user_id_idx ON system.users ("user_id" ASC)
+`
+
+func alterSystemUsersAddUserIDColumnWithIndex(
+	ctx context.Context, cs clusterversion.ClusterVersion, d upgrade.TenantDeps, _ *jobs.Job,
+) error {
+	for _, op := range []operation{
+		{
+			name:           "add-system-users-user-id-column",
+			schemaList:     []string{"user_id"},
+			query:          addUserIDColumn,
+			schemaExistsFn: hasColumn,
+		},
+		{
+			name:           "alter-system-users-add-index",
+			schemaList:     []string{"users_user_id_idx"},
+			query:          addUserIDIndex,
+			schemaExistsFn: hasIndex,
+		},
+	} {
+		if err := migrateTable(ctx, cs, d, op, keys.UsersTableID, systemschema.UsersTable); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func backfillSystemUsersIDColumn(
+	ctx context.Context, cs clusterversion.ClusterVersion, d upgrade.TenantDeps, _ *jobs.Job,
+) error {
+	var upsertRootStmt = `
+			       UPSERT INTO system.users (username, "isRole", "user_id") VALUES ($1, false, $2)
+			       `
+
+	var upsertAdminStmt = `
+			       UPSERT INTO system.users (username, "isRole", "user_id") VALUES ($1, true, $2)
+			       `
+
+	_, err := d.InternalExecutor.ExecEx(ctx, "upsert-root-user-in-role-id-migration", nil,
+		sessiondata.NodeUserSessionDataOverride, upsertRootStmt, username.RootUser, username.RootUserID)
+	if err != nil {
+		return err
+	}
+
+	_, err = d.InternalExecutor.ExecEx(ctx, "upsert-admin-role-in-role-id-migration", nil,
+		sessiondata.NodeUserSessionDataOverride, upsertAdminStmt, username.AdminRole, username.AdminRoleID)
+	if err != nil {
+		return err
+	}
+
+	row, err := d.InternalExecutor.QueryRowEx(ctx, `get-num-null-user-ids`, nil, sessiondata.NodeUserSessionDataOverride,
+		`SELECT count(1) FROM system.users WHERE user_id IS NULL`)
+	if err != nil {
+		return err
+	}
+	numUsersToUpdate := int(tree.MustBeDInt(row[0]))
+	const batchSize = 10000
+	for i := 0; i < numUsersToUpdate; i += batchSize {
+		numIDs := batchSize
+		if numUsersToUpdate-i < batchSize {
+			numIDs = numUsersToUpdate - i
+		}
+		startID, err := descidgen.IncrementUniqueRoleID(ctx, d.DB, d.Codec, int64(numIDs))
+		if err != nil {
+			return err
+		}
+
+		ids := make([]int, numIDs)
+		for j := range ids {
+			ids[j] = int(startID) + j
+		}
+
+		updateUserIDs := fmt.Sprintf(`
+UPDATE system.users
+   SET user_id = id
+  FROM (
+        SELECT *
+          FROM (
+                SELECT row_number() OVER () AS rn, *
+                  FROM system.users
+                 WHERE user_id IS NULL
+               ) AS t1
+               JOIN (
+                  SELECT row_number() OVER () AS rn,
+                         unnest AS id
+                    FROM ROWS FROM (unnest($1))
+                ) AS t2 ON t1.rn = t2.rn
+       ) AS temp
+ WHERE system.users.username = temp.username
+       AND system.users.user_id IS NULL
```

Comment:
How about the below command which also deals with updating the sequence?
```sql
  WITH start AS (
                SELECT max(user_id) AS start
                  FROM (
                         SELECT max(user_id) AS user_id FROM system.users
                         UNION ALL SELECT 100 -- the starting value for the sequence
                       )
             ),
       ids AS (
            SELECT i, i + start AS id
              FROM start, ROWS FROM (generate_series(0, 999)) AS t (i)
           ),
       users_to_update AS (
                        SELECT row_number() OVER () AS rn, name
                          FROM system.users
                         WHERE user_id IS NULL
                       ),
       updated AS (
                   UPDATE system.users
                      SET user_id = id
                     FROM (
                            SELECT name, id
                              FROM ids INNER JOIN users_to_update ON i = rn
                          )
                    WHERE username = name
                    LIMIT 1000
                RETURNING user_id
               )
SELECT setval('system.role_id_seq', max(user_id))
  FROM updated;
```

Reply:
I explored with the SQL statement above for a bit, one problem is that if there is a concurrent create user, the above as it as doesn't deal with it since it only sets the value at the end but doesn't block another concurrent create user from using it. Doing setval before doing updating works though. 

One more gripe from me is that the CTE isn't actually easier to understand compared to code, at least for me.

How much do you dislike it as it is?